File: app/src/main/java/com/metahumanz/pacilread/ui/CoverDecodeSizing.java
package com.metahumanz.pacilread.ui;

final class CoverDecodeSizing {
    private static final int DEFAULT_COVER_SIZE_PX = 360;

    private CoverDecodeSizing() {
    }

    static int sampleSizeFor(int width, int height, int targetWidth, int targetHeight) {
        int sampleSize = 1;
        int safeTargetWidth = Math.max(targetWidth, 1);
        int safeTargetHeight = Math.max(targetHeight, 1);
        while (width / (sampleSize * 2) >= safeTargetWidth
                && height / (sampleSize * 2) >= safeTargetHeight) {
            sampleSize *= 2;
        }
        return sampleSize;
    }

    static int targetSize(int measured) {
        return measured > 0 ? measured : DEFAULT_COVER_SIZE_PX;
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/ui/CoverDecodeSizing.kt
package com.metahumanz.pacilread.ui

internal object CoverDecodeSizing {
    private const val DEFAULT_COVER_SIZE_PX = 360

    @JvmStatic
    fun sampleSizeFor(width: Int, height: Int, targetWidth: Int, targetHeight: Int): Int {
        var sampleSize = 1
        val safeTargetWidth = targetWidth.coerceAtLeast(1)
        val safeTargetHeight = targetHeight.coerceAtLeast(1)
        while (width / (sampleSize * 2) >= safeTargetWidth &&
            height / (sampleSize * 2) >= safeTargetHeight
        ) {
            sampleSize *= 2
        }
        return sampleSize
    }

    @JvmStatic
    fun targetSize(measured: Int): Int = if (measured > 0) measured else DEFAULT_COVER_SIZE_PX
}
